gax/src/main/java/com/google/api/gax/rpc/StreamController.java

```diff
@@ -0,0 +1,90 @@
+/*
+ * Copyright 2017, Google LLC All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are
+ * met:
+ *
+ *     * Redistributions of source code must retain the above copyright
+ * notice, this list of conditions and the following disclaimer.
+ *     * Redistributions in binary form must reproduce the above
+ * copyright notice, this list of conditions and the following disclaimer
+ * in the documentation and/or other materials provided with the
+ * distribution.
+ *     * Neither the name of Google LLC nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package com.google.api.gax.rpc;
+
+import com.google.api.core.BetaApi;
+import java.util.concurrent.CancellationException;
+
+/**
+ * Allows the implementor of {@link ResponseObserver} to control flow of responses.
+ *
+ * <p>An instance of this class will be passed to {@link
+ * ResponseObserver#onStart(StreamController)}, at which point the receiver can disable automatic
+ * flow control. The receiver can also save a reference to the instance and terminate the stream
+ * early using {@code cancel()}.
+ */
+@BetaApi("The surface for streaming is not stable yet and may change in the future.")
+public abstract class StreamController {
+  /**
+   * Cancel the stream early.
+   *
+   * <p>This will manifest as an onError on the {@link ResponseObserver} with the cause being
+   * DEFAULT_CANCELLATION_EXCEPTION.
+   */
+  public final void cancel() {
+    cancel(new CancellationException("User cancelled stream"));
+  }
```

Comment:
If the only reason for an abstract class is to share this method, I would prefer an interface.

Reply:
Can you clarify what you mean? keep both cancels and make each subclass create its own default? remove one? or convert StreamController to an interface & create a AbstractStreamController that implements the cancel() method?

Reply:
I had in mind making each subclass create its own default.

Reply:
done